design/vela-core/environment.md

```diff
@@ -0,0 +1,292 @@
+# Managing Environments As Shared Bases For Application Deployment
+
+## Background
+
+A team of application development usually needs to prepare some shared environments for developers to deploy applications to.
+For example, most teams will have two environments, a *dev* environment for testing applications,
+and a *prod* environment for running applications to serve live traffic.
+The environment is a logical concept grouping common resources that multiple application deployments would depend on.
+
+A list of resources we want to define per environment:
+
+- K8s Clusters. Different environments might have clusters of different size and version, e.g.
+  small v1.10 k8s cluster for *dev* environment and large v1.9 k8s cluster for *prod* environment.
+- Admin policies. Production environment would usually set global policies on all application deployments,
+  like chaos testing, SLO requirements, security scanning, misconfiguration detection.
+- Operators & CRDs. Environments contains a variety of Operators & CRDs as system capabilities.
+  These operators can include domain registration, routing, logging, monitoring, autoscaling, etc.
+- Shared services. Environments contains a variety of system services that are be shared by applications.
+  These resources can include DB, cache, load balancers, API Gateways, etc.
+
+With the introduction of environment, the user workflow on Vela works like:
+
+- Platform team defines system Components for Environments.
+- Platform team deploys Environments using system Components.
+- Platform team defines user Components for Applications.
+- Developers choose user Components and Environments to deploy Applicatons.
+
+When deploying environments, there are additional requirements:
+
+- The ability to describe dependency relations between environments. For example, both *dev* and *prod* environments
+  would depend on a common environment that sets up operators both *dev* and *prod*  would use.
+
+
+## Environment CRD
+
+We propose to add an Environment CRD:
+
+```yaml
+kind: Environment
+metadata:
+  name: prod
+spec:
+  # An environment consists of multiple system resources.
+  # An environment is a grouping of these shared resources for user applications.
+  resources:
+    - template:
```

Comment:
why call it template?

We can category and distinguish here:

* `system`:  install only once, if the system already has one, no need to install again.
* `env`: install per env, always install, and will report an error if already exist.

Reply:
> env: install per env, always install, and will report an error if already exist.

I didn't catch this. Can you give an example?

Reply:
Note: we will add a field to indicate that this resource should be **owned** by this env.